rust/connlib/tunnel/src/control_protocol/client.rs

```diff
@@ -84,120 +91,88 @@ where
-                );
-
-                let peer = Arc::new(Peer::new(
-                    tunnel.private_key.clone(),
-                    index,
-                    peer_config.clone(),
-                    gateway_id,
-                    None,
-                    tunnel.rate_limiter.clone(),
-                ));
-
-                {
-                    let mut peers_by_ip = tunnel.peers_by_ip.write();
-
-                    for ip in peer_config.ips {
-                        peers_by_ip.insert(
-                            ip,
-                            ConnectedPeer {
-                                inner: peer.clone(),
-                                channel: d.clone(),
-                            },
-                        );
-                    }
-
-                    tunnel
-                        .role_state
-                        .lock()
-                        .gateway_awaiting_connection
-                        .remove(&gateway_id);
-                }
-
-                if let Some(conn) = tunnel.peer_connections.lock().get(&gateway_id) {
-                    conn.on_peer_connection_state_change(on_peer_connection_state_change_handler(
-                        gateway_id,
-                        tunnel.stop_peer_command_sender.clone(),
-                    ));
-                }
-
-                tokio::spawn(tunnel.start_peer_handler(peer, d));
-
-                tunnel
-                    .role_state
-                    .lock()
-                    .awaiting_connection
-                    .remove(&resource_id);
-            })
-        }));
+        let IceConnection {
+            ice_params,
+            ice_transport,
+            ice_candidate_rx,
+        } = new_ice_connection(&self.webrtc_api, relays).await?;
+        let preshared_key = self
+            .role_state
+            .lock()
+            .add_waiting_gateway(gateway_id, ice_candidate_rx);
+        self.peer_connections
+            .lock()
+            .insert(gateway_id, Arc::clone(&ice_transport));
 
-        let offer = peer_connection.create_offer(None).await?;
-        peer_connection.set_local_description(offer.clone()).await?;
+        set_connection_state_update(self, &ice_transport, gateway_id, resource_id);
 
         Ok(Request::NewConnection(RequestConnection {
             resource_id,
             gateway_id,
             client_preshared_key: Secret::new(Key(preshared_key.to_bytes())),
-            client_rtc_session_description: offer,
+            client_rtc_session_description: ice_params,
         }))
     }
 
+    fn new_tunnel(
+        &self,
+        resource_id: ResourceId,
+        gateway_id: GatewayId,
+        ice: Arc<RTCIceTransport>,
+    ) -> Result<()> {
+        let p_key = self
+            .role_state
+            .lock()
+            .gateway_preshared_keys
+            .get(&gateway_id)
+            .ok_or(Error::ControlProtocolError)?
+            .clone();
```

Comment:
We should be able to move this into `create_peer_config_for_new_connection` right?